### PR TITLE
- initial commit

### PR DESCRIPTION
**Pull request by pvpurcarcosmin**
refs: https://github.com/metasfresh/metasfresh/issues/7052

### CODE REVIEW

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentScheduleBL.java [hunk 1]
```diff
@@ -116,6 +44,102 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ShipmentScheduleBL.java:120]:
```suggestion
```

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java [hunk 2]
```diff
@@ -327,4 +346,59 @@ public I_M_ShipperTransportation getCommonM_ShipperTransportationOrNull(final Co
...
+	@NonNull
+	public ShipperTransportationId addTrackingCodesForInOutWithoutHU(@NonNull final AddTrackingCodesForInOutWithoutHUReq req)
+	{
+		final I_M_InOut inOut = inOutDAO.getById(req.getInOutId());
+
+		final WarehouseId warehouseId =  WarehouseId.ofRepoId(inOut.getM_Warehouse_ID());
+
+		final I_M_Warehouse warehouse = warehouseDAO.getById(warehouseId);
+
+		final int bPartnerLocationId = warehouse.getC_BPartner_Location_ID();
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java:365]: better call it `warehouseBPLocationId`

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java [hunk 3]
```diff
@@ -327,4 +346,59 @@ public I_M_ShipperTransportation getCommonM_ShipperTransportationOrNull(final Co
...
+
+		final BPartnerLocationInfo bPartnerLocationInfo = bPartnerLocationInfoRepository.getById(bPartnerLocationId);
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java:367]: better call it `warehouseBPLocationInfo`

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java [hunk 4]
```diff
@@ -327,4 +346,59 @@ public I_M_ShipperTransportation getCommonM_ShipperTransportationOrNull(final Co
...
+	{
+		final I_M_ShipperTransportation shipperTransportation = newInstance(I_M_ShipperTransportation.class);
+
+		shipperTransportation.setAD_Org_ID(request.getOrgId().getRepoId());
+		shipperTransportation.setM_Shipper_ID(request.getShipperId().getRepoId());
+		shipperTransportation.setShipper_BPartner_ID(request.getBPartnerLocationId().getBpartnerId().getRepoId());
+		shipperTransportation.setShipper_Location_ID(request.getBPartnerLocationId().getRepoId());
+		shipperTransportation.setDateDoc(Timestamp.valueOf(request.getShipDate()));
+
+		saveRecord(shipperTransportation);
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java:400]: introduce and use IShipperTransportationDAO.save(I_M_ShipperTransportation)

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java [hunk 5]
```diff
@@ -327,4 +346,59 @@ public I_M_ShipperTransportation getCommonM_ShipperTransportationOrNull(final Co
...
+
+	private ShipperTransportationId createHUShipperTransportation(@NonNull final CreateShipperTransportationRequest request)
+	{
+		final I_M_ShipperTransportation shipperTransportation = newInstance(I_M_ShipperTransportation.class);
+
+		shipperTransportation.setAD_Org_ID(request.getOrgId().getRepoId());
+		shipperTransportation.setM_Shipper_ID(request.getShipperId().getRepoId());
+		shipperTransportation.setShipper_BPartner_ID(request.getBPartnerLocationId().getBpartnerId().getRepoId());
+		shipperTransportation.setShipper_Location_ID(request.getBPartnerLocationId().getRepoId());
+		shipperTransportation.setDateDoc(Timestamp.valueOf(request.getShipDate()));
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java:398]: use `TimeUtil`

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/CreateShipperTransportationRequest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+public class CreateShipperTransportationRequest
+{
+	@NonNull
+	OrgId orgId;
+
+	@NonNull
+	ShipperId shipperId;
+
+	@NonNull
+	BPartnerLocationId bPartnerLocationId;
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/CreateShipperTransportationRequest.java:45]: better call it shipperBPartnerAndLocationId

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/CreateShipperTransportationRequest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import de.metas.shipping.ShipperId;
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Value;
+
+import java.time.LocalDateTime;
+
+@Value
+@Builder
+public class CreateShipperTransportationRequest
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/CreateShipperTransportationRequest.java:36]: consider changing to package level access

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/ApplyShipmentScheduleChangesRequest.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	BPartnerLocationId bPartnerLocationIdOverride;
+
+	@Nullable
+	BigDecimal qtyToDeliverStockingUOM;
+
+	@Nullable
+	List<CreateAttributeInstanceReq> attributes;
+
+	@Nullable
+	LocalDateTime deliveryDate;
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/ApplyShipmentScheduleChangesRequest.java:56]: why not ZonedDateTime?

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/AddTrackingCodesForInOutWithoutHUReq.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+public class AddTrackingCodesForInOutWithoutHUReq
+{
+	@NonNull
+	InOutId inOutId;
+
+	@NonNull
+	ShipperId shipperId;
+
+	@NonNull
+	LocalDateTime shipDate;
```
[the remark this answers is not in the document]
Comment (pvpurcarcosmin) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/AddTrackingCodesForInOutWithoutHUReq.java:45]: mainly bc there is no timezone info coming from Siro, but I guess I could manually set `Europe/Berlin` as zone id when reading the date from the Siro xmls, in order to have ZonedDateTime in our api.

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/CreateShipperTransportationRequest.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+	OrgId orgId;
+
+	@NonNull
+	ShipperId shipperId;
+
+	@NonNull
+	BPartnerLocationId bPartnerLocationId;
+
+	@NonNull
+	LocalDateTime shipDate;
```
[the remark this answers is not in the document]
Comment (pvpurcarcosmin) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/CreateShipperTransportationRequest.java:48]: same as above, will update

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java (hunk 4 above)
Comment (pvpurcarcosmin) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java:400]: why not ShipperTransportationRepository spring component?

Comment (teosarca) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java:400]: even better :)

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/shipping/ShipmentRestController.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+			shipmentService.validateRequest(request);
+		}
+		catch (final Exception e)
+		{
+			final String adLanguage = Env.getADLanguageOrBaseLanguage();
+			return ResponseEntity.badRequest()
+					.body(JsonErrors.ofThrowable(e, adLanguage));
+		}
+
+		request.getCreateShipmentInfoList()
```
[the remark this answers is not in the document]
Comment (pvpurcarcosmin) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/shipping/ShipmentRestController.java]: initially I had some logic for that, but after the latest changes it doesn't make sense anymore.

**Pull request merged**